### ISSUE TITLE
unknown array in the case of Container.Map matlab object

### ISSUE THREAD

**Issue by Amrit-Gill** [diffplug/matfilerw#7]
Following .mat file is a structure with a field called  'PMS' consisting a Container.Map object.
[ matlab.mat]( http://filebin.ca/2qyncp0fiBcF/matlab2.mat)

While reading, matfilerw is unable to recognize the dataType. Is it already a known issue.?

Thank you,
Amrit

**Comment by clokep** [comment]
I had done some work on this at some point and had partial support for it, `containers.Map` just got serialized as a normal Java object, you can check out [my implementation](https://github.com/clokep/jmatio/blob/80d9499735412464b035b499976730624a1d3277/JMatIO/prod/com/jmatio/types/MLJavaObject.java), which is probably incomplete (and probably depends on changes to `MLOpaque`).

**Comment by Amrit-Gill** [comment]
Thank you, I will test with your implementation and come back to you if there are any modifications.

**Comment by clokep** [comment]
It probably won't really work, that code is from a forked version of JMatIO, but it should get you part of the way there...

By the way, if anyone wants to include that in matfilerw, please feel free. Just include me as one of the authors. (Preferable with my newer email address...)

**Comment by nedtwigg** [comment]
Thanks for reporting this Amrit, especially the example!  It turns out that MatFileRW was already able to parse the MCOS data, but it was getting lost because MatFileRW only kept handles to "root" variables. Since your MCOS variable is a field, it was getting lost.  This has now been fixed in the latest snapshot version.  Here's some example code:

```java
MatFileReader mfr = new MatFileReader(new File("C:/Users/ntwigg/Downloads/matlab2.mat"));
MLStructure structure = (MLStructure) mfr.getContent().get("ans");
MLObjectPlaceholder pms = (MLObjectPlaceholder) structure.getField("PMS");
MLStructure serialization = (MLStructure) pms.getTarget().getField("serialization");
System.out.println("serialization=" + serialization.getFieldNames());
```

The output will be:

```java
serialization=[keys, values, uniformity, keyType, valueType]
```

Inside of each of those fields, there are cell arrays and structures full of what looks like useful data.  Feel free to reopen if you find further issues.

**Comment by nedtwigg** [comment]
Whoops, sorry, CI server found some bugs.  Standby...

**Comment by nedtwigg** [comment]
There we go, fixed, snapshot published, and much better too!  Here's the code you'd need now:

```java
MatFileReader mfr = new MatFileReader(new File("C:/Users/ntwigg/Downloads/matlab2.mat"));
MLStructure structure = (MLStructure) mfr.getContent().get("ans");
MLObject pms = (MLObjectPlaceholder) structure.getField("PMS");
MLStructure serialization = (MLStructure) pms.getField("serialization");
System.out.println("serialization=" + serialization.getFieldNames());
```

**Comment by nedtwigg** [comment]
Released in 2.3.0.
